File: clients/ios/Whisp/APIClient.swift
import Foundation
import Combine

class APIClient: ObservableObject {
    static let shared = APIClient()
    
    private let baseURL: String
    private var authToken: String?
    
    private init() {
        // Get API base URL from build configuration
        #if DEBUG
        // Development environment
        if let configURL = Bundle.main.object(forInfoDictionaryKey: "API_BASE_URL") as? String {
            self.baseURL = configURL
        } else {
            self.baseURL = "http://localhost:4000"
        }
        #else
        // Production environment
        if let configURL = Bundle.main.object(forInfoDictionaryKey: "API_BASE_URL") as? String {
            self.baseURL = configURL
        } else {
            self.baseURL = "https://api.whisp.app"
        }
        #endif
    }
    
    // MARK: - Authentication
    func setAuthToken(_ token: String) {
        authToken = token
    }
    
    func clearAuthToken() {
        authToken = nil
    }
    
    // MARK: - API Methods
    func register(username: String) async throws -> RegisterResponse {
        let url = URL(string: "\(baseURL)/register")!
        var request = URLRequest(url: url)
        request.httpMethod = "POST"
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")
        
        let body = RegisterRequest(username: username)
        request.httpBody = try JSONEncoder().encode(body)
        
        let (data, response) = try await URLSession.shared.data(for: request)
        
        guard let httpResponse = response as? HTTPURLResponse else {
            throw NetworkError.noData
        }
        
        switch httpResponse.statusCode {
        case 200...299:
            return try JSONDecoder().decode(RegisterResponse.self, from: data)
        case 409:
            throw NetworkError.conflict
        case 400:
            let error = try JSONDecoder().decode(APIError.self, from: data)
            throw NetworkError.serverError(error.error)
        default:
            throw NetworkError.serverError("Registration failed")
        }
    }
    
    func login(username: String, deviceToken: String) async throws -> LoginResponse {
        let url = URL(string: "\(baseURL)/login")!
        var request = URLRequest(url: url)
        request.httpMethod = "POST"
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")
        
        let body = LoginRequest(username: username, deviceToken: deviceToken)
        request.httpBody = try JSONEncoder().encode(body)
        
        let (data, response) = try await URLSession.shared.data(for: request)
        
        guard let httpResponse = response as? HTTPURLResponse else {
            throw NetworkError.noData
        }
        
        switch httpResponse.statusCode {
        case 200...299:
            return try JSONDecoder().decode(LoginResponse.self, from: data)
        case 401:
            throw NetworkError.unauthorized
        case 400:
            let error = try JSONDecoder().decode(APIError.self, from: data)
            throw NetworkError.serverError(error.error)
        default:
            throw NetworkError.serverError("Login failed")
        }
    }
    
    func addFriend(friendUsername: String) async throws {
        guard let token = authToken else {
            throw NetworkError.unauthorized
        }
        
        let url = URL(string: "\(baseURL)/friends/add")!
        var request = URLRequest(url: url)
        request.httpMethod = "POST"
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")
        request.setValue("Bearer \(token)", forHTTPHeaderField: "Authorization")
        
        let body = AddFriendRequest(friendUsername: friendUsername)
        request.httpBody = try JSONEncoder().encode(body)
        
        let (data, response) = try await URLSession.shared.data(for: request)
        
        guard let httpResponse = response as? HTTPURLResponse else {
            throw NetworkError.noData
        }
        
        switch httpResponse.statusCode {
        case 200...299:
            return
        case 401:
            throw NetworkError.unauthorized
        case 404:
            throw NetworkError.notFound
        case 400:
            let error = try JSONDecoder().decode(APIError.self, from: data)
            throw NetworkError.serverError(error.error)
        default:
            throw NetworkError.serverError("Add friend failed")
        }
    }
    
    func getFriends() async throws -> [Friend] {
        guard let token = authToken else {
            throw NetworkError.unauthorized
        }
        
        let url = URL(string: "\(baseURL)/friends")!
        var request = URLRequest(url: url)
        request.setValue("Bearer \(token)", forHTTPHeaderField: "Authorization")
        
        let (data, response) = try await URLSession.shared.data(for: request)
        
        guard let httpResponse = response as? HTTPURLResponse else {
            throw NetworkError.noData
        }
        
        switch httpResponse.statusCode {
        case 200...299:
            let friendsResponse = try JSONDecoder().decode(FriendsResponse.self, from: data)
            return friendsResponse.friends
        case 401:
            throw NetworkError.unauthorized
        default:
            throw NetworkError.serverError("Get friends failed")
        }
    }
    
    func getAllUsers() async throws -> [String] {
        let url = URL(string: "\(baseURL)/users")!
        let request = URLRequest(url: url)
        
        let (data, response) = try await URLSession.shared.data(for: request)
        
        guard let httpResponse = response as? HTTPURLResponse else {
            throw NetworkError.noData
        }
        
        switch httpResponse.statusCode {
        case 200...299:
            let usersResponse = try JSONDecoder().decode([String: [String]].self, from: data)
            return usersResponse["users"] ?? []
        default:
            throw NetworkError.serverError("Get users failed")
        }
    }
}

File: clients/android/app/src/main/java/com/rootkings/whisp/network/APIClient.kt
package com.rootkings.whisp.network

import android.content.Context
import com.rootkings.whisp.BuildConfig
import com.rootkings.whisp.models.*
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import okhttp3.OkHttpClient
import okhttp3.logging.HttpLoggingInterceptor
import retrofit2.Retrofit
import retrofit2.converter.gson.GsonConverterFactory
import retrofit2.http.*
import java.util.concurrent.TimeUnit

class APIClient(private val context: Context) {
    
    private val baseUrl: String
    private var authToken: String? = null
    
    init {
        // Get API base URL from build configuration
        baseUrl = if (BuildConfig.DEBUG) {
            // "http://10.0.2.2:4000" // Android emulator localhost
            "http://localhost:4000" // Use this to set IP of server
        } else {
            "https://api.whisp.app"
        }
    }
    
    private val okHttpClient = OkHttpClient.Builder()
        .addInterceptor { chain ->
            val request = chain.request().newBuilder()
            authToken?.let { token ->
                request.addHeader("Authorization", "Bearer $token")
            }
            chain.proceed(request.build())
        }
        .addInterceptor(HttpLoggingInterceptor().apply {
            level = if (BuildConfig.DEBUG) {
                HttpLoggingInterceptor.Level.BODY
            } else {
                HttpLoggingInterceptor.Level.NONE
            }
        })
        .connectTimeout(30, TimeUnit.SECONDS)
        .readTimeout(30, TimeUnit.SECONDS)
        .writeTimeout(30, TimeUnit.SECONDS)
        .build()
    
    private val retrofit = Retrofit.Builder()
        .baseUrl(baseUrl)
        .client(okHttpClient)
        .addConverterFactory(GsonConverterFactory.create())
        .build()
    
    private val apiService = retrofit.create(APIService::class.java)
    
    // MARK: - Authentication
    fun setAuthToken(token: String) {
        authToken = token
    }
    
    fun clearAuthToken() {
        authToken = null
    }
    
    // MARK: - API Methods
    suspend fun register(username: String): Result<RegisterResponse> = withContext(Dispatchers.IO) {
        try {
            val response = apiService.register(RegisterRequest(username))
            if (response.isSuccessful) {
                Result.success(response.body()!!)
            } else {
                val errorBody = response.errorBody()?.string()
                val error = if (errorBody != null) {
                    try {
                        val gson = com.google.gson.Gson()
                        val apiError = gson.fromJson(errorBody, APIError::class.java)
                        NetworkError.ServerError(apiError.error)
                    } catch (e: Exception) {
                        NetworkError.ServerError("Registration failed")
                    }
                } else {
                    NetworkError.ServerError("Registration failed")
                }
                Result.failure(error)
            }
        } catch (e: Exception) {
            Result.failure(NetworkError.ServerError(e.message ?: "Network error"))
        }
    }
    
    suspend fun login(username: String, deviceToken: String): Result<LoginResponse> = withContext(Dispatchers.IO) {
        try {
            val response = apiService.login(LoginRequest(username, deviceToken))
            if (response.isSuccessful) {
                Result.success(response.body()!!)
            } else {
                val error = when (response.code()) {
                    401 -> NetworkError.Unauthorized
                    400 -> {
                        val errorBody = response.errorBody()?.string()
                        val apiError = if (errorBody != null) {
                            try {
                                val gson = com.google.gson.Gson()
                                gson.fromJson(errorBody, APIError::class.java)
                            } catch (e: Exception) {
                                APIError("Login failed")
                            }
                        } else {
                            APIError("Login failed")
                        }
                        NetworkError.ServerError(apiError.error)
                    }
                    else -> NetworkError.ServerError("Login failed")
                }
                Result.failure(error)
            }
        } catch (e: Exception) {
            Result.failure(NetworkError.ServerError(e.message ?: "Network error"))
        }
    }
    
    suspend fun addFriend(friendUsername: String): Result<Unit> = withContext(Dispatchers.IO) {
        try {
            val response = apiService.addFriend(AddFriendRequest(friendUsername))
            if (response.isSuccessful) {
                Result.success(Unit)
            } else {
                val error = when (response.code()) {
                    401 -> NetworkError.Unauthorized
                    404 -> NetworkError.NotFound
                    400 -> {
                        val errorBody = response.errorBody()?.string()
                        val apiError = if (errorBody != null) {
                            try {
                                val gson = com.google.gson.Gson()
                                gson.fromJson(errorBody, APIError::class.java)
                            } catch (e: Exception) {
                                APIError("Add friend failed")
                            }
                        } else {
                            APIError("Add friend failed")
                        }
                        NetworkError.ServerError(apiError.error)
                    }
                    else -> NetworkError.ServerError("Add friend failed")
                }
                Result.failure(error)
            }
        } catch (e: Exception) {
            Result.failure(NetworkError.ServerError(e.message ?: "Network error"))
        }
    }
    
    suspend fun getFriends(): Result<List<Friend>> = withContext(Dispatchers.IO) {
        try {
            val response = apiService.getFriends()
            if (response.isSuccessful) {
                Result.success(response.body()!!.friends)
            } else {
                val error = when (response.code()) {
                    401 -> NetworkError.Unauthorized
                    else -> NetworkError.ServerError("Get friends failed")
                }
                Result.failure(error)
            }
        } catch (e: Exception) {
            Result.failure(NetworkError.ServerError(e.message ?: "Network error"))
        }
    }
    
    suspend fun getAllUsers(): Result<List<String>> = withContext(Dispatchers.IO) {
        try {
            val response = apiService.getAllUsers()
            if (response.isSuccessful) {
                val usersResponse = response.body()!!
                Result.success(usersResponse["users"] ?: emptyList())
            } else {
                Result.failure(NetworkError.ServerError("Get users failed"))
            }
        } catch (e: Exception) {
            Result.failure(NetworkError.ServerError(e.message ?: "Network error"))
        }
    }
}

// MARK: - API Service Interface
interface APIService {
    @POST("register")
    suspend fun register(@Body request: RegisterRequest): retrofit2.Response<RegisterResponse>
    
    @POST("login")
    suspend fun login(@Body request: LoginRequest): retrofit2.Response<LoginResponse>
    
    @POST("friends/add")
    suspend fun addFriend(@Body request: AddFriendRequest): retrofit2.Response<Unit>
    
    @GET("friends")
    suspend fun getFriends(): retrofit2.Response<FriendsResponse>
    
    @GET("users")
    suspend fun getAllUsers(): retrofit2.Response<Map<String, List<String>>>
}
